python/runtime/dbapi/paiio.py

```diff
@@ -0,0 +1,146 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License
+
+import re
+
+from runtime.dbapi.connection import Connection, ResultSet
+
+try:
+    import paiio
+except Exception:  # noqa: E722
+    pass
+
+
+class PaiIOResultSet(ResultSet):
+    def __init__(self, reader, err=None):
+        super().__init__()
+        self._reader = reader
+        self._column_info = None
+        self._err = err
+
+    def _fetch(self, fetch_size):
+        try:
+            return self._reader.read(num_records=fetch_size)
+        except Exception:  # noqa: E722
+            pass
+
+    def column_info(self):
+        """Get the result column meta, type in the meta maybe DB-specific
+
+        Returns:
+            A list of column metas, like [(field_a, INT), (field_b, STRING)]
+        """
+        if self._column_info is not None:
+            return self.column_info
+
+        schema = self._reader.get_schema()
+        columns = [(c['colname'], str.upper(c['typestr'])) for c in schema]
+        self._column_info = columns
+        return self._column_info
+
+    def success(self):
+        """Return True if the query is success"""
+        return self._reader is not None
+
+    def error(self):
+        return self._err
+
+    def close(self):
+        """
+        Close the ResultSet explicitly, release any resource incurred
+        by this query
+        """
+        if self._reader:
+            self._reader.close()
+            self._reader = None
+
+    def __del__(self):
+        self.close()
+
+
+class PaiIOConnection(Connection):
+    """PaiIOConnection emulate a connection for paiio,
+    currently only support full-table reading. That means
+    we can't filter the data, join the table and so on.
+    The only supported query statement is `None`.
+
+    Typical use is:
+    con = PaiIOConnection("paiio://db/tables/my_table")
+    res = con.query(None)
+    rows = [r for r in res]
+    """
+    def __init__(self, conn_uri):
+        super().__init__(conn_uri)
+        self.driver = "pai_maxcompute"
+        match = re.findall(r"paiio://\w+/tables/(.+)", self.uripts.path)
+        if len(match) < 1:
+            raise ValueError("Should specify table in uri with format: "
+                             "paiio://db/tables/table?param_a=a&param_b=b")
+        self.params["database"] = self.uripts.hostname
+        self.params["table"] = match[0]
+        self.params["slice_id"] = self.params.get("slice_id", 0)
+        self.params["slice_count"] = self.params.get("slice_count", 1)
+
+    def _get_result_set(self, statement):
+        if statement is not None:
+            raise ValueError("paiio only support full table read,"
+                             "so you need to pass statement with None.")
+        try:
+            reader = paiio.TableReader(self.params["table"],
```

Comment:
It is said that `paiio.TableRecordDataset` is faster than `TableReader`, and `TableReader` is actually not used currently.

Reply:
I will refer to the doc and change the function if possible.